### ISSUE TITLE
When downloading app file through 'http(s)://', the temporary zip file is never deleted

### ISSUE THREAD

**Issue by bgandon** [mevansam/terraform-provider-cf#86]
Hi,

To see this issue, here is the example resource I use:

```
resource "cf_app" "console_app" {
    name = "console"
    url = "https://github.com/orange-cloudfoundry/stratos-ui-cf-packager/releases/download/1.1.0/stratos-ui-packaged.zip"

    space = "${cf_space.ef_production_space.id}"
    route {
        default_route = "${cf_route.console_route.id}"
    }

    memory = 64 # MB
    disk_quota = 192 # MB
    timeout = 300 # seconds
    buildpack = "binary_buildpack"
    health_check_type = "port"

    depends_on = [ "cf_space.ef_production_space", "cf_route.console_route" ]
}
```

After a `terraform apply` the temporarily downloaded file is still there:

```
$ ls -ld /var/folders/j5/4jf9yvsj7w59djp8g__d13_w0000gn/T/*cfapp*
-rw-------  1 benjamin  staff  46699208 29 mai 19:35 /var/folders/j5/4jf9yvsj7w59djp8g__d13_w0000gn/T/cfapp137065753
```

And when testing it with `unzip -t`:

```
$ unzip -t /var/folders/j5/4jf9yvsj7w59djp8g__d13_w0000gn/T/cfapp137065753
Archive:  /var/folders/j5/4jf9yvsj7w59djp8g__d13_w0000gn/T/cfapp137065753
    ...
    testing: stratos-buildpack.yml    OK
    testing: stratos-dbmigrator       OK
    ...
    testing: user-info.so             OK
No errors detected in compressed data of /var/folders/j5/4jf9yvsj7w59djp8g__d13_w0000gn/T/cfapp137065753.
```

We can see that it's our `stratos-ui-packaged.zip` indeed.

The problem relies in the `prepareApp()` function in `resource_cf_app.go` because this function creates a temporary file and only returns its name. This makes no difference between a local Zip file to keep and a just-downloaded temporary file to delete. Once the `prepareApp()` function returns, there is no way for the calling code to know whether the file needs  to be deleted or not.

This issue is related to designing correct resource ownership in the code. It's always tricky to have a function that creates resources and let the responsibility for the calling code to release those resources.

The fix could be to return a structure that holds both the path to the file and whether it is a temporary file to be deleted when it is uploaded.

Best

**Comment by bgandon** [comment]
Oh, I realize this issue has been spotted in #38, and is currently being addressed in #58.

But the problem with #58 is that it hasn't showed any progress during the last month. Maybe a quick fix should be added to #70 or in a separate PR.

**Comment by janosbinder** [comment]
@bgandon Thanks for the note. We have fixed this issue in https://github.com/mevansam/terraform-provider-cf/pull/70 . If you still see this issue in the current `dev` edition of the plugin, please let us now.
